providers/default/macs/poly1305_prov.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/opensslconf.h>
+#ifndef OPENSSL_NO_POLY1305
+
+# include <openssl/core_numbers.h>
+# include <openssl/core_names.h>
+# include <openssl/params.h>
+# include <openssl/evp.h>
+# include <openssl/err.h>
+
+# include "internal/poly1305.h"
+# include "../../../crypto/poly1305/poly1305_local.h"
```

Comment:
Why is this necessary?

Reply:
Because line 42:

``` C
    POLY1305 poly1305;           /* Poly1305 data */
```

Reply:
You really should move the "internal" header file then.  Or turn the POLY135 and SIPHASH declarations into pointers.  Going up and over three levels is wrong: either keep the strutucre opaque or make it available internally.  Won't repeat the comment but it holds for siphash too.

Reply:
*Eventually*, I would like to see *all* of poly1305 move to the providers structure.  Alas, this is currently not possible, as there are legacy things in `crypto/poly1305` that we cannot remove, such as `crypto/poly1305/poly1305_ameth.c`, which is solely there to get MAC functionality via EVP_PKEY.

When *all* of poly1305 has moved to providers, I fully expect poly1305_local.h to move along with it, and that path will return to be just "poly1305_local.h".

I don't see the usefulness in moving this header file to something that is more "shared" (such as to `crypto/include/internal`).  I know that @slontis did this differently, but we disagree on how to do this, and that's ok.

Reply:
I thoroughly dislike seeing includes like this...and am not entirely convinced by the argument that "it will get moved eventually"...because such things have a habit of not actually happening and then they stick around for ever more. But I won't go so far as to insist on changing this.

Reply:
either way it needs a TODO on it.

Reply:
That is very true.  Will add that

Reply:
TODO added